File: apps/web/lib/loans.ts
// Loan applications client for the authenticated client dashboard.
//
// Calls GET /api/v1/loans/applications through the typed fetch wrapper in
// lib/api/client.ts. Wire shapes come from the generated contract
// (packages/contracts/generated/schema.d.ts); this module maps them to the
// camelCase shape the UI consumes, same pattern as getMe() in lib/auth.ts.

import type { components } from "@contracts/generated/schema";

import { apiRequest, type ApiResponse } from "@/lib/api/client";

type Schemas = components["schemas"];

export type LoanStatus = Schemas["LoanStatus"];
export type FeeOutcome = Schemas["FeeOutcome"];

export type LoanApplication = {
  id: string;
  loanTypeLabel: string;
  status: LoanStatus;
  statusReason: string | null;
  amountRequested: string | null;
  amountSanctioned: string | null;
  interestRate: string | null;
  processingFee: string | null;
  feeOutcome: FeeOutcome | null;
  openedOn: string; // ISO date
  closedOn: string | null;
  formVersion: number | null;
  formSchema: Schemas["ProductFormDefinition"] | null;
  formAnswers: Record<string, string | string[]> | null;
};

function mapApplication(raw: Schemas["LoanApplicationRead"]): LoanApplication {
  return {
    id: raw.id,
    loanTypeLabel: raw.loan_type.label,
    status: raw.status,
    statusReason: raw.status_reason,
    amountRequested: raw.amount_requested,
    amountSanctioned: raw.amount_sanctioned,
    interestRate: raw.interest_rate,
    processingFee: raw.processing_fee,
    feeOutcome: raw.fee_outcome,
    openedOn: raw.opened_at,
    closedOn: raw.closed_at,
    formVersion: raw.form_version,
    formSchema: raw.form_schema_snapshot,
    formAnswers: raw.form_answers,
  };
}

export async function getLoanApplications(): Promise<ApiResponse<LoanApplication[]>> {
  const res = await apiRequest<Schemas["LoanApplicationListResponse"]>("/api/v1/loans/applications");
  if (!res.ok) return res;
  return { ok: true, status: res.status, data: res.data.applications.map(mapApplication) };
}

export async function getLoanApplication(id: string): Promise<ApiResponse<LoanApplication>> {
  const res = await apiRequest<Schemas["LoanApplicationRead"]>(
    `/api/v1/loans/applications/${id}`,
  );
  if (!res.ok) return res;
  return { ok: true, status: res.status, data: mapApplication(res.data) };
}

export type FinancialProduct = Schemas["LoanTypeRead"];
export type LoanTypeOption = FinancialProduct;

export async function getLoanTypes(): Promise<ApiResponse<LoanTypeOption[]>> {
  const res = await apiRequest<Schemas["LoanTypeListResponse"]>("/api/v1/loans/loan-types");
  if (!res.ok) return res;
  return { ok: true, status: res.status, data: res.data.loan_types };
}

export type Bank = Schemas["BankRead"];

// loanTypeId filters to banks that offer that loan type (FR-6.3 per-bank
// availability) -- omitted, every active bank is returned unfiltered.
export async function getBanks(loanTypeId?: string): Promise<ApiResponse<Bank[]>> {
  const qs = loanTypeId ? `?loan_type_id=${loanTypeId}` : "";
  const res = await apiRequest<Schemas["BankListResponse"]>(`/api/v1/loans/banks${qs}`);
  if (!res.ok) return res;
  return { ok: true, status: res.status, data: res.data.banks };
}

export type LoanOfficerContact = {
  name: string;
  staffCode: string;
};

// Null is the "no officer assigned yet" state, not an error -- a client with
// no loan application yet, or one whose telecaller hasn't been assigned, has
// no officer to show. Contact routes through the support-ticket flow
// (lib/support-tickets.ts); this endpoint never returns phone/email.
export async function getMyLoanOfficer(): Promise<ApiResponse<LoanOfficerContact | null>> {
  const res = await apiRequest<Schemas["LoanOfficerContactRead"] | null>("/api/v1/loans/officer");
  if (!res.ok) return res;
  if (res.data === null) return { ok: true, status: res.status, data: null };
  return {
    ok: true,
    status: res.status,
    data: { name: res.data.name, staffCode: res.data.staff_code },
  };
}

export async function createLoanApplication(input: {
  productId: string;
  formVersion: number;
  answers: Record<string, string | string[]>;
}): Promise<ApiResponse<LoanApplication>> {
  const res = await apiRequest<Schemas["LoanApplicationRead"]>("/api/v1/loans/applications", {
    method: "POST",
    body: {
      loan_type_id: input.productId,
      form_version: input.formVersion,
      answers: input.answers,
    } satisfies Schemas["LoanApplicationCreate"],
  });
  if (!res.ok) return res;
  return { ok: true, status: res.status, data: mapApplication(res.data) };
}

export type FinancialServiceEnquiry = Schemas["FinancialServiceEnquiryRead"];

export async function getFinancialServiceEnquiries(): Promise<
  ApiResponse<FinancialServiceEnquiry[]>
> {
  const res = await apiRequest<Schemas["FinancialServiceEnquiryListResponse"]>(
    "/api/v1/loans/enquiries",
  );
  if (!res.ok) return res;
  return { ok: true, status: res.status, data: res.data.enquiries };
}

export async function createFinancialServiceEnquiry(input: {
  productId: string;
  formVersion: number;
  answers: Record<string, string | string[]>;
}): Promise<ApiResponse<FinancialServiceEnquiry>> {
  return apiRequest<FinancialServiceEnquiry>("/api/v1/loans/enquiries", {
    method: "POST",
    body: {
      product_id: input.productId,
      form_version: input.formVersion,
      answers: input.answers,
    } satisfies Schemas["FinancialServiceEnquiryCreate"],
  });
}

File: apps/api/app/schemas/loans.py
"""Client-facing loan schemas (/api/v1/loans/*).

Loan/product relationships are exposed as typed summaries rather than raw
foreign keys. Clients submit a published form version and dynamic answers;
identity, lead, business-line, workflow status, and canonical loan amount are
resolved or stamped server-side (see api/v1/loans.py).
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from app.models.loan import FeeOutcome, LoanStatus
from app.schemas.financial_products import FormAnswers, ProductCategory, ProductFormDefinition


class LoanTypeSummary(BaseModel):
    id: UUID
    label: str
    category: ProductCategory


class LoanApplicationRead(BaseModel):
    id: UUID
    loan_type: LoanTypeSummary
    status: LoanStatus
    status_reason: str | None
    amount_requested: Decimal | None
    amount_sanctioned: Decimal | None
    interest_rate: Decimal | None
    processing_fee: Decimal | None
    fee_outcome: FeeOutcome | None
    opened_at: datetime
    closed_at: datetime | None
    form_version: int | None
    form_schema_snapshot: ProductFormDefinition | None
    form_answers: FormAnswers | None


class LoanApplicationListResponse(BaseModel):
    applications: list[LoanApplicationRead]


class LoanApplicationCreate(BaseModel):
    loan_type_id: UUID
    form_version: Annotated[int, Field(ge=1)]
    answers: FormAnswers


class LoanTypeRead(BaseModel):
    id: UUID
    name: str
    label: str
    category: ProductCategory
    display_order: int
    form_version: int
    form_schema: ProductFormDefinition


class LoanTypeListResponse(BaseModel):
    loan_types: list[LoanTypeRead]


class FinancialServiceEnquiryCreate(BaseModel):
    product_id: UUID
    form_version: Annotated[int, Field(ge=1)]
    answers: FormAnswers


class FinancialServiceEnquiryRead(BaseModel):
    id: UUID
    product: LoanTypeSummary
    status: str
    form_version: int
    form_schema_snapshot: ProductFormDefinition
    form_answers: FormAnswers | None
    submitted_at: datetime


class FinancialServiceEnquiryListResponse(BaseModel):
    enquiries: list[FinancialServiceEnquiryRead]


class BankRead(BaseModel):
    id: UUID
    name: str


class BankListResponse(BaseModel):
    banks: list[BankRead]


class LoanOfficerContactRead(BaseModel):
    """Name + staff_code only -- never phone/email. Contact routes through
    the support-ticket flow (POST /api/v1/support-tickets/tickets)."""

    name: str
    staff_code: str


class LoanApplicationProgressUpdate(BaseModel):
    """Shared write shape for the Telecaller and Admin progression endpoints.

    Shape-only validation here (types/ranges); the transition/reason/terms-
    gating rules depend on the application's CURRENT status, so they live in
    services.loan_applications where that status is known.
    """

    status: LoanStatus | None = None
    status_reason: Annotated[str | None, Field(default=None, max_length=1000)] = None
    amount_sanctioned: Annotated[Decimal | None, Field(default=None, gt=0)] = None
    bank_id: UUID | None = None
    interest_rate: Annotated[Decimal | None, Field(default=None, ge=0, le=100)] = None
    processing_fee: Annotated[Decimal | None, Field(default=None, ge=0)] = None
    fee_outcome: FeeOutcome | None = None

    @model_validator(mode="after")
    def _at_least_one_field(self) -> LoanApplicationProgressUpdate:
        if all(
            v is None
            for v in (
                self.status,
                self.status_reason,
                self.amount_sanctioned,
                self.bank_id,
                self.interest_rate,
                self.processing_fee,
                self.fee_outcome,
            )
        ):
            raise ValueError("Provide at least one field to update.")
        return self
